### PR TITLE
Move all filtering to EntityList component | Closes #2517

### LINKED ISSUE #2517: Move all filtering to EntityList component
See [this comment on #2513](https://github.com/eventespresso/event-espresso-core/pull/2513/#discussion_r389211633)

### PR DESCRIPTION
**Pull request by manzoorwanijk**
This PR:
- Moves all the filtering, sorting, searching and pagination of entity lists to application
- Simplifies subscription types and adds more functions/features
- Creates filter bar service and updates UI Registry API
- Creates `useFilteredEntities` and `useEntityFilterService` to use in `EntityList`
- Creates `useDatesFilterBarService` and `useTicketsFilterBarService` to register the corresponding filters
- Creates domain `useFilterBarService` and hook` it up on EDTR init
- Updates `<DatesList />` and `<TicketsList />` to the new `<EntityList  />`
- Closes #2517

### CODE REVIEW

File: assets/src/application/services/subscription/useSubscriptionService.ts [hunk 1]
```diff
@@ -33,6 +25,21 @@ const useSubscriptionService: SubscriptionServiceHook = ({ domain, service }) =>
 		return window[NAMESPACE]?.[domain]?.[service]?.subscriptions || {};
 	};
 
+	const getServiceRegistryItem: SS['getServiceRegistryItem'] = (key) => {
+		return window[NAMESPACE]?.[domain]?.[service]?.[key];
```
Comment (nicolad) [inline assets/src/application/services/subscription/useSubscriptionService.ts]: this doesn't need a default value?

File: assets/src/application/ui/layout/entityList/filterBar/filterState/useEntityFilterService.ts [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+		let filteredEntities = entityList;
+
+		const callbacks = getCallbackList(mappedCallbackList);
+
+		callbacks.forEach((callback) => {
+			filteredEntities = callback({ entityList: filteredEntities, filterState });
+		});
+
+		return filteredEntities;
+	};
```
Comment (manzoorwanijk) [inline assets/src/application/ui/layout/entityList/filterBar/filterState/useEntityFilterService.ts]: I could not find anything in `Ramda` to handle this.

File: assets/src/application/ui/layout/entityList/filterBar/filterState/useEntityFilterService.ts [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+
+type ELFSM = EntityListFilterStateManager;
+
+const useEntityFilterService = <D extends string, L extends string, E extends Entity, FS extends ELFSM>(
+	domain: D,
+	listId: L
+): EntityFilterService<E, FS> => {
+	type EFS = EntityFilterService<E, FS>;
+
+	const { getFilters, getSearches, getSorters } = useFilterBarService<D, L, E, ELFSM>(domain, listId);
```
Comment (nicolad) [inline assets/src/application/ui/layout/entityList/filterBar/filterState/useEntityFilterService.ts]: interesting service

File: assets/src/application/services/subscription/useSubscriptionService.ts (hunk 1 above)
Comment (manzoorwanijk) [reply assets/src/application/services/subscription/useSubscriptionService.ts]: No, the consumer needs to handle that.

Comment (tn3rb) [inline assets/src/application/services/subscription/useSubscriptionService.ts]:
```suggestion
	const getServiceRegistryItem: SS['getServiceRegistryItem'] = (key, default) => {
		return window[NAMESPACE]?.[domain]?.[service]?.[key] || default;
```

File: assets/src/application/services/subscription/useSubscriptionService.ts [hunk 4]
```diff
@@ -33,6 +25,21 @@ const useSubscriptionService: SubscriptionServiceHook = ({ domain, service }) =>
 		return window[NAMESPACE]?.[domain]?.[service]?.subscriptions || {};
```
Comment (tn3rb) [inline assets/src/application/services/subscription/useSubscriptionService.ts]:
```suggestion
		return getServiceRegistryItem('subscriptions, {}');
```

only works if a default param is added to that function though

File: assets/src/application/ui/layout/entityList/filterBar/subscription/useFilterBarService.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+
+	const getCallbacks = ({ listId = entityListId, type }) => {
+		invariant(listId, 'No `listId` provided');
+		const allSubscriptions = getSubscriptions();
+
+		const isForList = pathEq(['options', 'listId'], listId);
+		const isOfType = pathEq(['options', 'type'], type);
+		const isOfTypeAndForList = allPass([isForList, isOfType]);
+
+		return filter(isOfTypeAndForList, allSubscriptions);
```
Comment (nicolad) [inline assets/src/application/ui/layout/entityList/filterBar/subscription/useFilterBarService.ts]: :+1:

File: assets/src/application/ui/layout/entityList/filterBar/filterState/useFilteredEntities.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	}, [total, searchResults]);
+
+	useEffect(() => {
+		// If there are no paginated entities and current pageNumber is not 1
+		//e.g. When there is only one entity on the last page and it's deleted
+		if (paginatedEntities.length === 0 && pageNumber > 1) {
+			// goto the previous page
+			setPageNumber(pageNumber - 1);
+		}
+	}, [paginatedEntities.length]);
```
Comment (tn3rb) [inline assets/src/application/ui/layout/entityList/filterBar/filterState/useFilteredEntities.ts]: should this also use `perPage` && `pageNumber` as dependencies?

File: assets/src/application/ui/layout/entityList/filterBar/subscription/useFilterBarService.ts (hunk 5 above)
Comment (tn3rb) [inline assets/src/application/ui/layout/entityList/filterBar/subscription/useFilterBarService.ts]: nice

File: assets/src/application/services/subscription/useSubscriptionService.ts (hunk 4 above)
Comment (manzoorwanijk) [reply assets/src/application/services/subscription/useSubscriptionService.ts]: This actually triggers some TS errors which would complicate the things further, that's why I left it as such

File: assets/src/application/ui/layout/entityList/filterBar/filterState/useFilteredEntities.ts (hunk 6 above)
Comment (manzoorwanijk) [reply assets/src/application/ui/layout/entityList/filterBar/filterState/useFilteredEntities.ts]: No, that's only needed when an entity is deleted.

**Pull request merged**